internal/controller/vrg_volgrouprep.go

```diff
@@ -311,14 +311,184 @@ func (v *VRGInstance) getVGRCFromVGR(vgr *volrep.VolumeGroupReplication) (volrep
+func (v *VRGInstance) getVGRUsingSCLabel(pvc *corev1.PersistentVolumeClaim) (*volrep.VolumeGroupReplication, error) {
+	rID, err := v.getVGRClassReplicationID(pvc)
+	if err != nil {
+		// Error is masked here, as caller expects k8serrors regarding vgr resource
+		return nil, fmt.Errorf("error determining replicationID")
+	}
+
+	vgrNamespacedName := types.NamespacedName{
+		Name:      rmnutil.TrimToK8sResourceNameLength(rID + v.instance.Name),
+		Namespace: pvc.Namespace,
+	}
+
+	vgr := &volrep.VolumeGroupReplication{}
+	err = v.reconciler.Get(v.ctx, vgrNamespacedName, vgr)
+
+	return vgr, err
+}
+
+// getVGRClassReplicationID is a utilti function that fetches the replicationID for the PVC looking at the class labels
+func (v *VRGInstance) getVGRClassReplicationID(pvc *corev1.PersistentVolumeClaim) (string, error) {
+	pvcNamespacedName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
+
+	vgrClass, err := v.selectVolumeReplicationClass(pvcNamespacedName, true)
+	if err != nil {
+		return "", err
+	}
+
+	replicationID, ok := vgrClass.GetLabels()[VolumeReplicationIDLabel]
+	if !ok {
+		v.log.Info(fmt.Sprintf("VolumeGroupReplicationClass %s is missing replicationID for PVC %s/%s",
+			vgrClass.GetName(), pvc.GetNamespace(), pvc.GetName()))
+
+		return "", fmt.Errorf("volumeGroupReplicationClass %s is missing replicationID for PVC %s/%s",
+			vgrClass.GetName(), pvc.GetNamespace(), pvc.GetName())
+	}
+
+	return replicationID, nil
+}
+
+// ensurePVCUnprotected returns true if the passed in PVC is not protected by the vgr
+func (v *VRGInstance) ensurePVCUnprotected(
+	pvc *corev1.PersistentVolumeClaim,
+	vgr *volrep.VolumeGroupReplication,
+) bool {
+	const unprotected = true
+
+	if rmnutil.ResourceIsDeleted(vgr) {
+		return !unprotected
+	}
+
+	for i := range vgr.Status.PersistentVolumeClaimsRefList {
+		if vgr.Status.PersistentVolumeClaimsRefList[i].Name == pvc.GetName() {
+			return !unprotected
+		}
+	}
+
+	return unprotected
+}
+
+// deleteVGRIfUnused garbage collects a VGR that is not protecting any PVCs
+func (v *VRGInstance) deleteVGRIfUnused(vgr *volrep.VolumeGroupReplication) error {
+	if len(vgr.Status.PersistentVolumeClaimsRefList) != 0 {
+		return nil
+	}
+
+	err := v.reconciler.Delete(v.ctx, vgr)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+	// TODO: Delete VGR from S3 store
+
+	return nil
+}
+
 //nolint:gocognit
 func (v *VRGInstance) pvcsUnprotectVolGroupRep(groupPVCs map[types.NamespacedName][]*corev1.PersistentVolumeClaim) {
+	// Single PVC that is deselected/deleted will not come here, in a circutious way!
+	// VRG deletion will invoke this routine
 	for vgrNamespacedName, pvcs := range groupPVCs {
 		log := v.log.WithValues("vgr", vgrNamespacedName.String())
 
 		vgrMissing, requeueResult := v.reconcileMissingVGR(vgrNamespacedName, pvcs, log)
 		if vgrMissing || requeueResult {
-			v.requeue()
+			if requeueResult {
+				v.requeue()
+			}
```

Comment:
If vgrMissing alone does not trigger v.requeue(), what is the reason to include it in the if condition?